capsules/src/print_log.rs

```diff
@@ -0,0 +1,276 @@
+        // processes with things to write by looking at the
+        // sequence number; if the next-to-be-given sequence
+        // number != this process+1, then a process has
+        // grabbed this process+1 and has something to send.
+        self.alarm
+            .set_alarm(self.alarm.now(), self.alarm.ticks_from_ms(10));
+    }
+}
+
+impl<'a, A: Alarm<'a>> AlarmClient for PrintLog<'a, A> {
+    fn alarm(&self) {
+        //debug!("Alarm fired.");
+        if self.tx_in_progress.get() {
+            self.tx_in_progress.set(false);
+
+            // Issue an upcall for the in-progress write
+            for cntr in self.apps.iter() {
+                cntr.enter(|app, kernel_data| {
+                    // This is the in-progress write
+                    if app.writing {
+                        let _res = kernel_data.schedule_upcall(1, (app.write_len, 0, 0));
+                        app.writing = false;
+                    }
+                });
+            }
+        }
+
+        // Find if there's another writer and mark it busy.
+        let mut next_writer: Option<ProcessId> = None;
+        let mut seqno = self.tx_counter.get();
+
+        // Find the process that has an outstanding write with the
+        // lowest sequence number.
+        for cntr in self.apps.iter() {
+            let appid = cntr.processid();
+            cntr.enter(|app, _| {
+                if app.pending_write {
+                    // This means app.tx_counter is smaller than seqno, as long as
+                    // there aren't usize/2 processes.
+                    if seqno.wrapping_sub(app.tx_counter) < usize::MAX / 2 {
+                        seqno = app.tx_counter;
+                        next_writer = Some(appid);
+                    }
+                }
+            });
+        }
+
+        next_writer.map(|pid| {
+            self.apps.enter(pid, |app, kernel_data| {
+                app.pending_write = false;
+                self.send(app, kernel_data);
+            })
+        });
+    }
+}
+
+impl<'a, A: Alarm<'a>> SyscallDriver for PrintLog<'a, A> {
+    /// Setup shared buffers.
+    ///
+    /// ### `allow_num`
+    ///
+
+    /// Setup shared buffers.
+    ///
+    /// ### `allow_num`
+    ///
+    /// - `0`: Readonly buffer for write buffer
+
+    // Setup callbacks.
+    //
+    // ### `subscribe_num`
+    //
+    // - `1`: Write buffer completed callback
+
+    /// Initiate serial transfers
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver check.
+    /// - `1`: Transmits a buffer passed via `allow`, up to the length
+    ///        passed in `arg1`
+    fn command(&self, cmd_num: usize, arg1: usize, _: usize, appid: ProcessId) -> CommandReturn {
+        //debug!("PrintLog receiving command: {} {} from {:?}", cmd_num, arg1, appid);
+        let res = self
+            .apps
+            .enter(appid, |app, kernel_data| {
+                match cmd_num {
+                    0 => Ok(()),
```

Comment:
I don't think this really matters for this capsule, but I don't think we want anyone to copy this example, as command(0) should still return Ok() even if something went wrong with creating the grant.

Reply:
This was copied from console; I was trying to keep the same behavior. I can change both in a separate PR?